doc/developer-guide/internal-libraries/ArgParser.en.rst

```diff
@@ -0,0 +1,304 @@
+.. Licensed to the Apache Software Foundation (ASF) under one
+   or more contributor license agreements.  See the NOTICE file
+   distributed with this work for additional information
+   regarding copyright ownership.  The ASF licenses this file
+   to you under the Apache License, Version 2.0 (the
+   "License"); you may not use this file except in compliance
+   with the License.  You may obtain a copy of the License at
+
+   http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing,
+   software distributed under the License is distributed on an
+   "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+   KIND, either express or implied.  See the License for the
+   specific language governing permissions and limitations
+   under the License.
+
+.. include:: ../../common.defs
+
+.. highlight:: cpp
+.. default-domain:: cpp
+
+.. _ArgParser:
+
+ArgParser
+*********
+
+Synopsis
+++++++++
+
+.. code-block:: cpp
+
+   #include <ts/ArgParser.h>
+
+Description
++++++++++++
+
+:class:`ArgParser` is a powerful and easy-to-use command line Parsing library for ATS.
+The program defines what it requires by adding commands and options.
+Then :class:`ArgParser` will figure out what is related from the command line.
+All parsed arguments and function will be put in a key-value pairs structure
+:class:`Arguments` available for users to use.
+
+Usage
++++++
+
+The usage of the ArgParser is straightforward. The user is expected to create an
+ArgParser for the program. Commands and options can be added to the parser with details
+including ENV variable, arguments expected, etc. After a single method :code:`parse(argv)` is called,
+An object containing all information and parsed arguments available to use will be returned.
+
+Create a parser
+---------------
+The first step to use argparse is to create a parser.
+The parser can be created simply by calling:
+
+.. code-block:: cpp
+
+   ts::ArgParser parser;
+
+or initialize with the following arguments: 
+*name, help description, environment variable, argument number expected, function*
+
+.. code-block:: cpp
+
+   ts::ArgParser parser("name", "description", "ENV_VAR", 0, &function);
+
+To add the usage for the help message of this program:
+
+.. code-block:: cpp
+
+   parser.add_global_usage("traffic_blabla [--SWITCH]");
+
+Add commands and options
+------------------------
+
+We can perform all kinds of operations on the parser which is the :class:`ArgParser` or command which is a :class:`Command`.
+
+To add commands to the program or current command:
+
+.. code-block:: cpp
+
+   parser.add_command("command", "description");
+   command1.add_command("command", "description");
```

Comment:
Where did `command1` come from?

Reply:
That is just an example. Maybe I'll change it a bit to make it easier to understand.

Reply:
Examples should be better than code. They must never be hard to understand, it defeats the entire purpose of the example. Note that you can directly include source code using `literalinclude`. I use this to pull in example code from the unit tests, which is nice because then I know the example actually compiles and runs.